lib/bot_framework/session.rb

```diff
@@ -0,0 +1,422 @@
+module BotFramework
+  class Session #< Events::EventEmitter
+    attr_accessor :library, :message, :user_data, :conversation_data, :private_conversation_data,
+                  :session_state, :dialog_state, :localizer
+    def initialize(options)
+      super
+      @library = options[:library]
+      @localizer = options[:localizer]
+      @msg_sent = false
+      @is_reset = false
+      @last_send_time = Time.now # TODO: Move to proc?
+      @batch = []
+      @batch_timer = Timers::Group.new
+      @batch_started = false
+      @sending_batch = false
+      @in_middleware = false
+      @_locale = nil
+
+      @options = options || {
+        on_save: nil,
+        on_send: nil,
+        library: nil,
+        localizer: nil,
+        middleware: nil,
+        dialog_id: nil,
+        dialog_args: nil,
+        auto_batch_delay: nil,
+        dialog_error_message: nil,
+        actions: nil
+      }
+      @auto_batch_delay = 250 unless options[:auto_batch_delay].is_a? Integer
+      @timers = Timers::Group.new
+    end
+
+    def to_recognize_context
+      {
+        message: message,
+        user_data: user_data,
+        conversation_data: conversation_data,
+        private_conversation_data: private_conversation_data,
+        localizer: localizer,
+        dialog_stack: dialog_stack,
+        preferred_locale: preferred_locale,
+        get_text: get_text,
+        nget_text: nget_text,
+        locale: preferred_locale
+      }
+    end
+
+    def dispatch(session_state, message)
+      index = 0
+      session = self
+      now = Time.now
+      middleware = @options[:middleware] || []
+
+      _next = lambda do
+        handler = middleware[index] if index < middleware.length
+        if handler
+          index += 1
+          handler(session, _next)
+        else
+          @in_middleware = false
+          @session_state[:last_acess] = now
+          done
+        end
+      end
+      session_state ||= { call_stack: [], last_acess: Time.now, version: 0.0 }
+      # Making sure that dialog is properly initialized
+      cur = cur_dialog
+      self.dialog_data = cur.state if cur
+      # Dispatch message
+      message ||= { text: '' }
+      message[:type] ||= 'message'
+      # Ensure that localized prompts are loaded
+      # TODO
+      self
+    end
+
+    def error error
+      logger.info "Session error"
+      if options[:dialog_error_message] 
+        end_conversation(options[:dialog_error_message])
+      else
+        #TODO Add localisation
```

Comment:
Annotation keywords like `TODO` should be all upper case, followed by a colon, and a space, then a note describing the problem.